lib/Transforms/SetSPIRVAbiAttribute.cpp

```diff
@@ -16,40 +16,75 @@
 
 #include <imex/Transforms/Passes.h>
 
+#include "llvm/ADT/SmallVector.h"
 #include <mlir/Dialect/GPU/Transforms/Passes.h>
 #include <mlir/Dialect/MemRef/IR/MemRef.h>
 #include <mlir/Dialect/SPIRV/IR/SPIRVDialect.h>
 #include <mlir/Dialect/SPIRV/IR/TargetAndABI.h>
+#include <mlir/Pass/Pass.h>
 
 namespace imex {
-struct SetSPIRVAbiAttribute
-    : public mlir::PassWrapper<SetSPIRVAbiAttribute,
-                               mlir::OperationPass<mlir::gpu::GPUModuleOp>> {
-
-  virtual void
-  getDependentDialects(mlir::DialectRegistry &registry) const override {
-    registry.insert<mlir::gpu::GPUDialect>();
-    registry.insert<mlir::spirv::SPIRVDialect>();
+#define GEN_PASS_DEF_SETSPIRVABIATTRIBUTE
+#include "imex/Transforms/Passes.h.inc"
+} // namespace imex
+
+namespace {
+class SetSPIRVAbiAttributePass
+    : public imex::impl::SetSPIRVAbiAttributeBase<SetSPIRVAbiAttributePass> {
+public:
+  explicit SetSPIRVAbiAttributePass() { m_clientAPI = "vulkan"; }
+  explicit SetSPIRVAbiAttributePass(const mlir::StringRef &clientAPI)
+      : m_clientAPI(clientAPI) {}
+
+  mlir::LogicalResult initializeOptions(mlir::StringRef options) override {
+    if (failed(Pass::initializeOptions(options)))
+      return mlir::failure();
+
+    if (clientAPI == "opencl") {
+      m_clientAPI = "opencl";
+    }
+
+    if (clientAPI != "vulkan" && clientAPI != "opencl")
+      return mlir::failure();
+
+    return mlir::success();
   }
 
   void runOnOperation() override {
     auto gpuModule = getOperation();
     auto *context = &getContext();
     auto attrName =
         mlir::StringAttr::get(context, mlir::spirv::getEntryPointABIAttrName());
-    auto abi = mlir::spirv::getEntryPointABIAttr(llvm::None, context);
-    for (auto gpuFunc : gpuModule.getOps<mlir::gpu::GPUFuncOp>()) {
-      if (!mlir::gpu::GPUDialect::isKernel(gpuFunc) ||
-          gpuFunc->getAttr(attrName))
-        continue;
+    if (m_clientAPI == "opencl") {
```

Comment:
it seems only getEntryPointABIAttr argument differs between opencl and vulkan, can this code be restructured to avoid code duplication?